src/framework/MOM_io.F90

```diff
@@ -887,6 +890,60 @@ subroutine read_variable_1d_int(filename, varname, var, ncid_in)
   call broadcast(var, size(var), blocking=.true.)
 end subroutine read_variable_1d_int
 
+!> Read a 2d array from a netCDF input file and save to a variable.
+!!
+!! Start and nread ranks may exceed var, but must match the rank of the
+!! variable in the netCDF file.  This allows for reading slices of larger
+!! arrays.
+!!
+!! I/O occurs only on the root PE, and data is broadcast to other ranks.
+!! Due to potentially large memory communication and storage, this subroutine
+!! should only be used when domain-decomposition is unavaialable.
+subroutine read_variable_2d(filename, varname, var, start, nread, ncid_in)
+  character(len=*), intent(in) :: filename  !< Name of file to be read
+  character(len=*), intent(in) :: varname   !< Name of variable to be read
+  real, intent(out)            :: var(:,:)  !< Output array of variable
+  integer, optional, intent(in) :: start(:) !< Starting index on each axis.
+  integer, optional, intent(in) :: nread(:) !< Number of values to be read along each axis
+  integer, optional, intent(in) :: ncid_in  !< netCDF ID of an opened file.
+              !! If absent, the file is opened and closed within this routine.
+
+  integer :: ncid, varid, ndims, rc
+  character(len=*), parameter :: hdr = "read_variable_2d"
+  character(len=128) :: msg
+
+  if (is_root_pe()) then
+    if (present(ncid_in)) then
+      ncid = ncid_in
+    else
+      call open_file_to_Read(filename, ncid)
+    endif
+
+    call get_varid(varname, ncid, filename, varid, match_case=.false.)
+    if (varid < 0) call MOM_error(FATAL, "Unable to get netCDF varid for "//trim(varname)//&
+                                         " in "//trim(filename))
+    ! Verify that start(:) and nread(:) ranks match variable's dimension count
+    rc = nf90_inquire_variable(ncid, varid, ndims=ndims)
+    if (rc /= NF90_NOERR) call MOM_error(FATAL, hdr // trim(nf90_strerror(rc)) //&
+          " Difficulties reading "//trim(varname)//" from "//trim(filename))
+
+    ! NOTE: We could check additional information here (type, size, ...)
+    rc = nf90_get_var(ncid, varid, var, start, nread)
+    if (rc /= NF90_NOERR) call MOM_error(FATAL, hdr // trim(nf90_strerror(rc)) //&
+          " Difficulties reading "//trim(varname)//" from "//trim(filename))
+
+    if (size(start) /= ndims .or. size(nread) /= ndims) then
```

Comment:
I think that there a several problems with this test.

1. Because the `start` and `nread` arguments are optional, there needs to be test for whether they are present to wrap the test for whether they have the right size.  Otherwise, this code  as written could lead to a segmentation fault.
2. This error handling block should probably be moved up 6 lines, so that it occurs before `start` and `nread` are used in the call to nf90_get_var.
3. The fact that this code appears to work, suggests that perhaps start and nread are not optional after all?
4. `ndims` is a property of an input file, whereas the sizes of `start` and `nread` are effectively hard coded in the call to read_variable. Could there be a case where it would be appropriate to set start and nread so that this routine could, for example, read either from a 2-d file or read a single level from a 3-d file into a 2-d array, depending on what is in the file that is being read?  If so, issuing an fatal error might not be the right answer.

  Could this work properly if the tests were changed to `if (present(start) .and. present(nread)) then ; if ((size(start) < ndims) .or. size(nread) < ndims) then`, and then the call to nf90_get_var were changed to
 ` if (present(start) .and. present(nread) then
    call nf90_get_var(ncid, varid, var, start(1:ndims), nread(1:ndims))
  else
    call nf90_get_var(ncid, varid, var)
  endif`
```

Reply:
Optional arguments are pass-through, so the call to `nf90_get_var` does not need to be changed.  But I agree with the general conclusions here.